### ISSUE TITLE
Swift: let or var for viewModel properties

### ISSUE THREAD

**Issue by klaaspieter** [thoughtbot/guides#435]
There's been some discussion lately about using `let` or `var` for view models. On [Markable](https://github.com/instantag/markable-vip-ios/pull/103/commits/81c5c02d52633577d11ca7148ae717c757cd0340#diff-7ea7710b47b0d7b31ed42537c7a16ab1R66) we've been using `var`. I'm not sure which one of this is better but I feel we should have a place to discuss this where the conversations doesn't get lost. 

I like `var` because it's practical. We can update the view model when needed without having to recreate the view model or abstract that process away into a Lens or something similar.

I like `let` because it feels more pure. Using `var` means everything is mutable so we need to ensure we keep track of state. @tonyd256 has been advocating `let` actively so I think it's better if he explains his position and how he's been using view models.

**Comment by tonyd256** [comment]
I think `var` has a particular advantage in a situation (that I'll explain in a bit) that `let` does not. So some might say, "Why not always use `var` then?". I think the answer safety. So now let's get into it.

When I have an "MVVM" architecture I like having a View/ViewController (V/VC), Controller (C), ViewModel (VM), and a Model (M) for each "screen" (this can vary). The V/VC has an associated C that vends the VM to the V/VC so it can display itself.

So how does this VM get created? I've used 2 different methods to do this and I think both have their merits.

1: The VM is a `var` on the C and each time that the data changes, like via a request or sort/filter function, the function that changes the data also initializes a new VM and sets the `var`. One can then use the `didSet` attribute on the VM `var` to notify its observers to redraw.

```swift
final class Controller {
  weak var observer: ViewModelObserver?
  var results = []

  var viewModel = ViewModel(d1: results, d2: .None, d3: "") {
    didSet {
      observer?.viewModelDidUpdate()
    }
  }

  func getInitialValues() {
    makeNetworkRequest { results in
      self.results = results

      self.viewModel = ViewModel(d1: results, d2: self.viewModel.d2, d3: self.viewModel.d3)
      // or with `var`
      self.viewModel.d1 = results
    }
  }

  func filter(text: String) {
    viewModel = ViewModel(results: results.filter(text), text: text)
    // or with `var`
    viewModel.results = results.filter(text)
    viewModel.text = text
  }
}
```

2: The VM is a computed property on the C that initializes the VM when requested. This means that the C has it's own references to the data so the VM can use those to initialize itself. Now the data modification functions set the C's datums. One can then tell the observers to update in the data modification functions or in the `didSet`s on the individual datums.

```swift
final class Controller {
  weak var observer: ViewModelObserver?
  var results = []
  var filteredResults = []
  var filterText = ""

  var viewModel: ViewModel {
    return ViewModel(results: filteredResults, text: filterText)
  }

  func getInitialValues() {
    makeNetworkRequest { results in
      self.results = results
      self.filteredResults = results
      observers?.viewModelDidUpdate()
    }
  }

  func filter(text: String) {
    filterText = text
    filteredResults = results.filter(text)
    observers?.viewModelDidUpdate()
  }
}
```

I think `1` has a definite advantage when using `var` in the VM; however, I much prefer the `2` pattern. It feels better to me because of less direct VM modification. I'd be curious how others would weigh in between these 2. Please consider more than just total lines and drawback that are in both.

**Comment by tonyd256** [comment]
One thing I'm not sure about is if setting one property on a VM will trigger the `didSet` on the C's VM variable. Can anyone confirm?

**Comment by sharplet** [comment]
@tonyd256 yep nesting shouldn't affect property observers of everything's a struct. So for example `view.frame.size.height = newHeight` would trigger a property observer for `view.frame`, and `self.viewModel.results.append(fetchedResult)` would trigger `viewModel`.

**Comment by dazmuda** [comment]
I think `2` is only viable if your viewModels are flimsy little model wrappers with formatters and convenience methods, since it's a computed property. Which is what they should be.

`2` also seems advantageous when  the controller needs to query the model. For example, makeNetworkRequest(objectID) requires either `controller.model.id` or `controller.viewModel.model.id`. There's no reason the controller should have to parley with the viewModel for a model property that is required for controller activities.

Overall `2` aligns better with my idea of a viewModel's role. Sitting in-between the view and the model, not living on a controller and holding onto the model for the controller.

**Comment by tonyd256** [comment]
I realized I forgot to explain why I said this:
>I think the answer safety.

So I think it comes down to being able to change the VM in multiple places. It's kinda like multiple returns to me where the more you have the more there is a chance for error.

Anyways, I also want to mention that everything I've said has focused on VMs and not the Models (M) that also can have `var` or `let`. This is also part of this discussion and I don't think I have as good as an answer as I have for the VM case outlined above.

I think, in general, I would default to `let` and use `var` if I found it helped me out a lot. I think in some cases we can work around the `var` helpfulness for a better overall experience like case `2` of the VM architecture, but there will definitely be time when `var` will me more useful. I've started to look at all the things we have in Swift as tools and not worrying too much about which is worse than another but more about which is better in different scenarios. Use our entire tool box to make good code work good.
